lib/IDE/CompletionInstance.cpp

```diff
@@ -162,10 +165,79 @@ static DeclContext *getEquivalentDeclContextFromSourceFile(DeclContext *DC,
   return newDC;
 }
 
+/// Check if any dependent files are modified since \p timestamp.
+bool areAnyDependentFilesInvalidated(CompilerInstance &CI,
+                                     llvm::vfs::FileSystem &FS,
+                                     StringRef currentFileName,
+                                     llvm::sys::TimePoint<> timestamp) {
+
+  auto isInvalidated = [&](StringRef filePath) -> bool {
+    llvm::errs() << "filePath: " << filePath << "\n";
+    auto stat = FS.status(filePath);
+    if (!stat)
+      // Missing.
+      return true;
+
+    auto lastModTime = stat->getLastModificationTime();
+    if (lastModTime > timestamp)
+      // Modified.
+      return true;
+
+    // If the last modification time is zero, this file is probably from a
+    // virtual file system. We need to check the content.
+    if (lastModTime == llvm::sys::TimePoint<>()) {
+      if (&CI.getFileSystem() == &FS)
```

Comment:
> Are these ever going to be equal, or in practice we'll always get a new VFS object?

I don't think so. I believe this is unreachable.

> Even if the pointers are equal, is it correct to assume the contents did not change? The VFS object could be mutable.

You are right. We perhaps should retain all buffers of the VFS dependencies. But I think it's good enough for now because the VFS objects should be always different.

Reply:
I would say keep a hash not the whole buffer.

Reply:
OK. I updated the PR to keep hashes of the dependencies.